### ISSUE TITLE
String Literal Type can't be used as key

### ISSUE THREAD

**Issue by olostan** [Microsoft/TypeScript#14032]
**TypeScript Version:**  2.1.3 

**Code**
```ts
type X = "a" | "b";

const x: { [key: X]: string } = {
    "a":"value"
}
```

**Expected behavior:**
Should compile as X is actually a string. 

**Actual behavior:**
Compile-time error:
```
An index signature parameter type must be 'string' or 'number'.
```

**Comment by gcnew** [comment]
Use the `in` operator. E.g.

```ts
type X = "a" | "b";

const zzz: { [key in X]: string } = {
    "a": "value",
    "b": "value 2"     // b is required
}

// to make `b` optional make the keys optional
const zzz2: { [key in X]?: string } = {
    "a": "value",
}
```

**Comment by masaeedu** [comment]
@gcnew Neither of those is quite the same as requiring a single key from the union.

**Comment by gcnew** [comment]
@masaeedu Yes, that's true. Unfortunately picking a single key is not currently possible, AFAIK :(

**Comment by masaeedu** [comment]
@gcnew I am looking to construct the type `type A = { foo: 10 } | { bar: 10 }`. Using the literal types `type B = "foo" | "bar"` and `type C = 10`. I feebly remember recognizing some general distributive property of unions over other types in a discussion about invoking unions of functions, but I can't quite figure it out for key-value map types.

At the moment there is no value-level computation involved, this is purely type-level juggling of expressions, so I don't understand what you're saying in the edit. Could you provide an example of what you mean?

**Comment by masaeedu** [comment]
@olostan It turns out this is possible, you just need to introduce a type variable:

```
type X = "a" | "b";
function acceptX<T extends X>(x: {[K in T]: string }) {
    return x
}
const x = acceptX({ "b": "value" })
// x is inferred as { "b": string }
```

Passing `{ "c": "value" }` or `{ "a": 10 }` is a type error, as expected.
